runtime/commonMain/src/kotlinx/serialization/modules/SerialModuleBuilders.kt

```diff
@@ -148,15 +149,28 @@ public class SerializersModuleBuilder internal constructor() : SerialModuleColle
         }
     }
 
+    @JvmName("registerSerializer") // Don't mangle method name for prettier stack traces
     internal fun <T : Any> registerSerializer(
         forClass: KClass<T>,
         serializer: KSerializer<T>,
         allowOverwrite: Boolean = false
     ) {
-        if (!allowOverwrite && forClass in class2Serializer) throw SerializerAlreadyRegisteredException(forClass)
+        if (!allowOverwrite) {
+            val previous = class2Serializer[forClass]
+            if (previous != null && previous != serializer) {
+                // TODO when working on SD rework, provide a way to properly stringify serializer as its FQN
+                val currentName = serializer.descriptor.serialName
+                val previousName = previous.descriptor.serialName
+                throw SerializerAlreadyRegisteredException(
+                    "Serializer for $forClass already registered in this module: $previous ($previousName), " +
+                            "attempted to register $serializer ($currentName)"
+                )
+            }
+        }
         class2Serializer[forClass] = serializer
     }
 
+    @JvmName("registerPolymorphicSerializer") // Don't mangle
```

Comment:
[nit] you not always read file from top to bottom continuously, so I'd suggest copy-paste comment `// Don't mangle method name for prettier stack traces` here completely